clients/adapter-amqp/src/main/java/org/eclipse/hono/adapter/client/command/amqp/ProtonBasedAdapterInstanceCommandHandler.java

```diff
@@ -70,38 +68,28 @@ public void handleCommandMessage(final Message msg, final ProtonDelivery deliver
-                : ResourceIdentifier.fromString(msg.getAddress());
-        if (resourceIdentifier == null || resourceIdentifier.getResourceId() == null) {
+
+        final ProtonBasedCommand command;
+        try {
+            command = ProtonBasedCommand.fromRoutedCommandMessage(msg);
+        } catch (final IllegalArgumentException e) {
             LOG.debug("address of command message is invalid: {}", msg.getAddress());
             final Rejected rejected = new Rejected();
             rejected.setError(new ErrorCondition(Constants.AMQP_BAD_REQUEST, "invalid command target address"));
             delivery.disposition(rejected, true);
             return;
         }
-        final String tenantId = resourceIdentifier.getTenantId();
-        final String originalDeviceId = resourceIdentifier.getResourceId();
-        // fetch "via" property (if set)
-        final String gatewayIdFromMessage = MessageHelper.getApplicationProperty(msg.getApplicationProperties(),
-                MessageHelper.APP_PROPERTY_CMD_VIA, String.class);
-        final String targetDeviceId = gatewayIdFromMessage != null ? gatewayIdFromMessage : originalDeviceId;
-        final CommandHandlerWrapper commandHandler = commandHandlers.getCommandHandler(tenantId, targetDeviceId);
-
-        // Adopt gateway id from command handler if set;
-        // for that kind of command handler (gateway subscribing for specific device commands), the
-        // gateway information is not stored in the device connection service ("deviceConnectionService.setCommandHandlingAdapterInstance()" doesn't have an extra gateway id parameter);
-        // and therefore not set in the delegated command message
-        final String gatewayId = commandHandler != null && commandHandler.getGatewayId() != null
-                ? commandHandler.getGatewayId()
-                : gatewayIdFromMessage;
-
-        final ProtonBasedCommand command = new ProtonBasedCommand(msg, tenantId,
-                gatewayId != null ? gatewayId : originalDeviceId);
+        final CommandHandlerWrapper commandHandler = commandHandlers.getCommandHandler(command.getTenant(),
+                command.getGatewayOrDeviceId());
+        if (commandHandler != null && commandHandler.getGatewayId() != null) {
+            // Gateway information set in command handler means a gateway has subscribed for commands for a specific device.
+            // This information isn't getting set in the message (by the Command Router) and therefore has to be adopted manually here.
+            command.setGatewayId(commandHandler.getGatewayId());
+        }
 
         final SpanContext spanContext = TracingHelper.extractSpanContext(tracer, msg);
-        final Span currentSpan = org.eclipse.hono.client.impl.CommandConsumer.createSpan("handle command", tenantId,
-                originalDeviceId, gatewayId, tracer, spanContext);
+        final Span currentSpan = org.eclipse.hono.client.impl.CommandConsumer.createSpan("handle command",
```

Comment:
I assume this dependency will be removed in an additional PR?

Reply:
Yes.